### ISSUE TITLE
Expecting existing DebugElement toBeFalsy creates circular reference in Jasmine

### ISSUE THREAD

**Issue by AmyBlankenship** [angular/angular#14235]
**I'm submitting a ...**  (check one with "x")
```
[X ] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
When you test to see if a DebugElement that exists but shouldn't is falsey, Jasmine churns on Plunker but eventually returns sort of. On Angular-CLI on WebStorm, it will keep cycling with a circular reference as it tries to print out the DebugElement to compare it to falsey. It's actually hard to figure out that's what happens because it just scrolls and scrolls and the only chance you have to Ctrl-C is when karma is restarting after it gets fed up.

**Expected behavior**
String representation of debug element should be able to be written to console in a reasonable amount of time.

**Minimal reproduction of the problem with instructions**
https://plnkr.co/edit/6a7iKt80GAVdhW9MRrPx

**What is the motivation / use case for changing the behavior?**
People who embrace test first development will write the minimal template without the *ng directive that hides it under testing circumstances. 

**Please tell us about your environment:**
Windows 10, Webstorm 2016.2.4 Angular-CLI 1.0.0-beta.20-4

* **Angular version:** 2.0.X
Locally I have 2.1.0, but I'm assuming that since I forked your plunk for the steps to reproduce, it's still an issue.

* **Browser:** Edge, but I imagine it's all

 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TS

* **Node (for AoT issues):** `node --version` =

**Comment by PanolifeHB** [comment]
I experienced a similar issue when testing an existing DebugElement to not exist. 

```
let deleteButtonElement = fixture.debugElement.query(By.css('.delete-button'));
expect(testTour).toBeFalsy('Delete Button should not exist');
```

It shows the same behaviour no matter what I tried to test for: toBeFalsy, toBeNull, toBe(null). Karma just disconnects at a certain point because of a timeout without failing the test case:

>Disconnect (1 times), because no message in 100000ms

Unfortunately it also does not really give you a clue, which test case is responsible for the timeout. This might be really annoying when a test case fails months later due to some refactoring...

Is there maybe any other way for testing whether an element exists or not that does not cause this problem?

**Comment by DzmitryShylovich** [comment]
`expect(testTour != null).toBe(true)`?

**Comment by PanolifeHB** [comment]
Thanks, that would be a good workaround for the issue.

**Comment by AmyBlankenship** [comment]
At the very least, this should be called out in the docs. It should not be left up to each developer to discover this issue and figure out how to fix it.

**Comment by DzmitryShylovich** [comment]
You should also be able to set `jasmine.MAX_PRETTY_PRINT_DEPTH` to something lower than the default of 40

**Comment by AmyBlankenship** [comment]
Doesn't Angular 2 overwrite some Jasmine functionality? Could this be part of that overwrite?

**Comment by tbosch** [comment]
Background: `DebugElement` contains an instance of `ViewData`, which is just a plain old JavaScript object. Jasmine tries to pretty print `DebugElement` including the `ViewData`. However, `ViewData` and the related `ViewDefinition` is non trivial...

**Comment by Blasz** [comment]
A PR has been submitted for this #15074. Can we add the bug type label to this and assign some people to review the PR? It was created some time ago and is a trivial fix.

**Comment by tbosch** [comment]
Hi,
this is not the right fix though, as the underlying data structure in Angular is used in multiple projects. The better way of doing this is to set `jasmine.pp` to a function that pretty prints only to a certain depth / size of plain old javascript objects.

**Comment by ilikeprograms** [comment]
Thank you guys so much. I have been writing unit tests for hours and not knowing why this has been timing out and theres no output.

@DzmitryShylovich I added MAX_PRETTY_PRINT_DEPTH and I suddenly got told why my test was failing.

**Comment by psxvoid** [comment]
It's still an issue in angular 4.4.4

![screenclip](https://user-images.githubusercontent.com/11395921/31228218-5a8b0e2a-a9e5-11e7-85d6-0cf03a9d767f.png)

**Comment by seteen** [comment]
I encountered this bug too.

**Comment by kpfromer** [comment]
I have encountered this bug as well.

**Comment by dwilches** [comment]
I hit this bug today in Angular 4.2.4.
While this issue gets addressed, I posted a workaround here:
https://stackoverflow.com/questions/44301315/karma-error-no-captured-browser-open-http-localhost9876/48606192

**Comment by etovian** [comment]
Bump!  This just cost me about 30 minutes.  Thanks for the workaround!  And to restate what @AmyBlankenship mentioned about TDD, once you've got a directive in place to keep the element from being rendered (eg, *ngIf), then expect(el).toBeFalsy() works just fine.  This problem only occurs when you start with a failing test that expects the element to be falsy before the directive is in place.

**Comment by brendoncaulkins** [comment]
Ran into this yesterday, spent half a day working on it with a teammate.  Came at it this morning by redlining the code, and found the `fixture.debugElement.query()` line to be the point that hangs (not the expect).  That led me here and to the accompanying bug in jasmine (https://github.com/jasmine/jasmine/issues/1263)

Angular: 6.1.10
Jasmine: 2.8.0

Code that was hanging:
```
const checkBox = fixture.debugElement.query(By.css('#recurring-source-0'))
expect(checkBox).toBeNull()
```

Workaround:
```
const checkBoxes = fixture.debugElement.queryAll(By.css('#recurring-source-0'))
expect(checkBoxes.length).toBe(0)
```

**Comment by fetis** [comment]
even there's a workaround, I don't see it's very usable. You have to repeat it in every test you check the DOM. Not a clean solution, imo.

**Comment by JoostK** [comment]
There have been two updates regarding this bug:

1. Jasmine's pretty printer has been adjusted to prevent serializing objects that would overflow the max string length in https://github.com/jasmine/jasmine/commit/c0d051319990858de86c46a9697711bf80a2ef7c. Released in Jasmine 2.9.0 (the repo case here is using Jasmine 2.4.1)
1. The `DebugElement` of Ivy only refers to the HTML node, not any internal Angular datastructures.

Given the above advancements, there's no more actions to be done for this issue report.
